src/logic/map_objects/tribes/soldier.cc

```diff
@@ -1603,6 +1602,116 @@ void Soldier::die_pop(Game& game, State& /* state */) {
 	schedule_destroy(game);
 }
 
+Bob::Task const Soldier::taskNavalInvasion = {
+   "naval_invasion", static_cast<Bob::Ptr>(&Soldier::naval_invasion_update), nullptr, nullptr,
+   true};
+
+void Soldier::start_task_naval_invasion(Game& game, const Coords& coords) {
+	push_task(game, taskNavalInvasion);
+	State& invasion_state = top_state();
+
+	invasion_state.coords = coords;
+	invasion_state.ivar1 = 0;
+	invasion_state.ivar2 =
+	   game.descriptions().get_building_descr(owner().tribe().port())->get_conquers();
+	invasion_state.ivar3 = game.descriptions().get_largest_workarea();
+
+	molog(game.get_gametime(), "[naval_invasion] Starting at %3dx%3d\n", coords.x, coords.y);
+}
+
+void Soldier::naval_invasion_update(Game& game, State& state) {
+	std::string signal = get_signal();
+	if (!signal.empty()) {
+		signal_handled();
+	}
+
+	/*
+	 * Soldiers in a naval invasion are based around `state.coords` and seek to eliminate
+	 * all enemy influence within a radius of `state.ivar2 + state.ivar3` (where ivar2 is the
+	 * conquer radius of our planned port and ivar3 the biggest possible enemy conquer radius).
+	 * When such influence is found, the target building serial is stored in `state.ivar1`.
+	 * Otherwise, they stay at the flag and wait indefinitely until we get around
+	 * to implementing the code for the next step.
+	 */
+
+	if (state.ivar1 != 0) {
+		// Heading to enemy building
+		Building* enemy = dynamic_cast<Building*>(game.objects().get_object(state.ivar1));
+		state.ivar1 = 0;
+
+		if (enemy == nullptr || !owner().is_hostile(enemy->owner()) ||
+		    enemy->attack_target() == nullptr || !enemy->attack_target()->can_be_attacked()) {
+			molog(game.get_gametime(), "[naval_invasion] Attack target vanished or not hostile\n");
+			return schedule_act(game, Duration(10));
+		}
+
+		start_task_attack(game, *enemy, CF_NONE);
+		return schedule_act(game, Duration(10));
+	}
+
+	// No enemy, look for one
+	const Map& map = game.map();
+
+	std::vector<ImmovableFound> results;
+	map.find_reachable_immovables(game, Area<FCoords>(get_position(), state.ivar2 + state.ivar3),
```

Comment:
Looking for reachable attack targets is actually pretty efficient, I did not notice any FPS drop when running this with an invasion force of 30 soldiers. Hollow circles would not gain anything because of the check for reachability as the soldiers may need to meander around obstacles anyway.

Due to deterministic sorting order, soldiers will at first all one-by-one start attacking the same building *until* so many of them are heading there that subsequent soldiers cannot find a clear path to the target and pick the next-nearest building, and so on.